iothub/device/src/Pipeline/RetryDelegatingHandler.cs

```diff
@@ -1090,12 +1091,41 @@ protected override void Dispose(bool disposing)
                     {
                         _handleDisconnectCts?.Cancel();
                         _handleDisconnectCts?.Dispose();
-                        if (_handlerSemaphore != null && _handlerSemaphore.CurrentCount == 0)
+
+                        if (_clientOpenCloseSemaphore != null && _clientOpenCloseSemaphore.CurrentCount == 0)
+                        {
+                            _clientOpenCloseSemaphore.Release();
+                        }
+                        _clientOpenCloseSemaphore?.Dispose();
+                        _clientOpenCloseSemaphore = null;
```

Comment:
Maybe we shouldn't assign to null so we get an object disposed exception instead of NRE.

Reply:
I like that, sure. We're not relying on this field being null to indicate any special scenario, so I see no issues in us disposing this semaphore and then leaving it to the garbage collector to eventually clear up the references.